### PR TITLE
workflows: publish approved PRs on CI completion

### PR DESCRIPTION
**Pull request by carlocab**
This is a continuation of #128697 and #128544.

We need to now check the PR's `mergeable_state` to make sure we're merging something with a green approval (rather than a grey one).

See octokit/octokit.net#1763 for more details on `mergeable_state`.

### CODE REVIEW

File: .github/workflows/publish-commit-bottles.yml [hunk 1]
```diff
@@ -137,6 +139,13 @@ jobs:
             exit 1
           fi
 
+          # See https://github.com/octokit/octokit.net/issues/1763 for possible values.
+          if [[ "$mergeable_state" != "clean" ]] && [[ "$mergeable_state" != "unstable" ]]
+          then
+            echo "::error ::PR #$PR mergeable_state is $mergeable_state!"
+            exit 1
+          fi
+
```
Comment (ZhongRuoyu) [inline .github/workflows/publish-commit-bottles.yml:148]: I think we'll need a retry mechanism here if `$mergeable_state` is `unknown`.

```console
$ for i in {0..3}; do gh api repos/Homebrew/homebrew-core/pulls/123791 -q .mergeable_state; sleep 1; done
unknown
blocked
blocked
blocked
```

Also, do we want to check for the PR's draft status?

File: .github/workflows/automerge.yml
Comment (ZhongRuoyu) [inline .github/workflows/automerge.yml:1]: Do we want to check for the PR's draft status here too?

Comment (carlocab) [reply .github/workflows/automerge.yml:1]: Yes, done.

File: .github/workflows/publish-commit-bottles.yml (hunk 1 above)
Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml]: I moved the check into `automerge.yml` which already has a retry mechanism for querying the API.

**Comment by ZhongRuoyu** [comment]
Sorry, didn't intend to click that. (I was using the mobile app.)

File: .github/workflows/automerge.yml [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            fi
+
+            echo "::notice ::PR #$PR status:"
+            echo "::notice ::Approved?    $approved"
+            echo "::notice ::CI Complete? $complete"
+            echo "::notice ::Mergeable?   $mergeable"
+            echo "::notice ::Checking again in ${timeout}s..."
+            sleep "$timeout"
+            attempt=$(( attempt + 1 ))
+            timeout=$(( timeout * 2 ))
```
[the remark this answers is not in the document]
Comment (carlocab) [reply .github/workflows/automerge.yml:143]: Shortened initial timeout.

File: .github/workflows/automerge.yml [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+          do
+            if [[ "$approved" = "false" ]]
+            then
+              approved="$(
+                gh api \
+                  --header 'Accept: application/vnd.github+json' \
+                  --header 'X-GitHub-Api-Version: 2022-11-28' \
+                  "repos/$GH_REPO/pulls/$PR/reviews" \
+                  --jq 'any(.[].state; . == "APPROVED")'
+              )"
```
[the remark this answers is not in the document]
Comment (carlocab) [reply .github/workflows/automerge.yml]: There is some overlap (also with doing `gh pr checks`), but there's more documentation about the `state` attribute of each review, so I'd rather rely on that where possible.

File: .github/workflows/automerge.yml [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR: ${{ steps.pr.outputs.number }}
+        run: |
+          attempt=0
+          max_attempts=10
+          timeout=2
+
+          approved=false
+          complete=false
+          mergeable=false
```
Comment (ZhongRuoyu) [inline .github/workflows/automerge.yml:88]: Shoule these be put inside the loop so that they are re-checked in each iteration? In case they are changed.

**Review by MikeMcQuaid — DISMISSED** [review]
Looking good so far!

File: .github/workflows/automerge.yml (hunk 4 above)
Comment (carlocab) [reply .github/workflows/automerge.yml:88]: We can, but I'd start to be wary about rate limits this way.

Comment (ZhongRuoyu) [reply .github/workflows/automerge.yml:88]: Right, we have a limit of 1,000 requests per hour...

Then perhaps another option would be to rely on the "Check PR approval" step in the `publish-commit-bottles` workflow to have a final check. Does it sound necessary to check `mergeable_state` and draft status there too?

Sorry for being nitpicking 😅 -- I realised that I'm basically picturing something very abnormal. If that sounds overcautious, I'm also fine with how it is right now.

Comment (carlocab) [reply .github/workflows/automerge.yml]: Bumped the number of attempts down. I _think_ this should be safe for the rate limit? If not we can switch it to using a PAT I guess.

Comment (ZhongRuoyu) [reply .github/workflows/automerge.yml]: Sounds okay to me. We'll see!

**Review by ZhongRuoyu — APPROVED** [review]
Thanks for your great work!

**Comment by carlocab** [comment]
For the record, if anything 💥, it's easy enough to disable the automerge workflow without having to do any reverts:

Screenshot 2023-04-21 at 00 11 06

**Comment by ZhongRuoyu** [comment]
We're seeing a number of failures, because the (outdated) CI workflow did not produce the `pull-number` artifact. But that's expected I guess.

**Pull request merged**